src/components/WalletHistory.vue

```diff
@@ -47,89 +47,142 @@
+  @Getter externalActivity!: AccountHistory<HistoryItem>;
+  @Getter externalActivityTotal!: number;
   @Getter shouldBalanceBeHidden!: boolean;
+  @Action getAssets!: AsyncVoidFn;
   @Action navigate!: (options: { name: string; params?: object }) => Promise<void>;
+  @Action getExternalActivity!: (options?: any) => Promise<void>;
+  @Action resetExternalActivity!: AsyncVoidFn;
   @Action getAccountActivity!: AsyncVoidFn;
+  @Action clearSyncedAccountActivity!: (options: { address: string; assetAddress?: string }) => Promise<void>;
 
   @Prop() readonly asset?: AccountAsset;
 
-  TransactionStatus = TransactionStatus;
-  query = '';
-  currentPage = 1;
-  pageAmount = 8;
-
-  // for fast search
-  get activityHashTable(): any {
-    return this.activity.reduce((result, item) => {
-      if (!item.txId) return result;
-      return { ...result, [item.txId]: item };
-    }, {});
+  @Watch('searchQuery')
+  private async updateHistoryBySearchQuery() {
+    await this.updateCommonHistory();
   }
 
-  get assetAddress(): string | undefined {
-    return this.asset && this.asset.address;
+  readonly pageAmount = 8; // override PaginationSearchMixin
+  readonly updateCommonHistory = debounce(() => this.updateHistory(true, true), 500);
+
+  get assetAddress(): string {
+    return (this.asset && this.asset.address) || '';
   }
 
-  get transactions(): Array<History> {
-    if (this.assetAddress) {
-      return this.activity.filter((item) => [item.assetAddress, item.asset2Address].includes(this.assetAddress));
-    }
-    return this.activity;
+  get internalHistory(): Array<History> {
+    const activityList = Object.values(this.activity);
+    const prefiltered = this.assetAddress
+      ? activityList.filter((item) => {
+          return [item.assetAddress, item.asset2Address].includes(this.assetAddress);
+        })
+      : activityList;
+
+    return prefiltered;
   }
 
-  get filteredHistory(): Array<any> {
-    if (!this.hasHistory) return [];
-    return this.getFilteredHistory(this.transactions).sort((a: History, b: History) =>
-      a.startTime && b.startTime ? b.startTime - a.startTime : 0
-    );
+  get filteredInternalHistory(): Array<History> {
+    return this.getFilteredHistory(this.internalHistory);
   }
 
-  get hasHistory(): boolean {
-    return !!(this.transactions && this.transactions.length);
+  get filteredExternalHistory(): Array<History> {
+    return Object.values(this.externalActivity);
+  }
+
+  get transactions(): Array<History> {
+    const merged = [...this.filteredInternalHistory, ...this.filteredExternalHistory];
+    const sorted = this.sortTransactions(merged);
+
+    return this.getPageItems(sorted);
   }
 
   get total(): number {
-    return this.filteredHistory.length;
+    return this.externalActivityTotal + this.internalHistory.length;
+  }
+
+  get hasTransactions(): boolean {
+    return !!(this.transactions && this.transactions.length) || !!this.searchQuery;
```

Comment:
We have `transactions.length ` 3 times in this file. Should we have variable for it?